### PR TITLE
Clean up quantifiers engine initialization.

### PR DESCRIPTION
**Pull request by ajreynol**
This includes using unique_ptr uniformly for all modules and utilities.  It also includes some work towards #1163.

### CODE REVIEW

**Comment by ajreynol** [comment]
@timothy-king , this is partially motivated by coverity following your results from #2356.

File: src/theory/quantifiers_engine.cpp [hunk 1]
```diff
@@ -89,41 +115,32 @@ QuantifiersEngine::QuantifiersEngine(context::Context* c,
...
-  d_util.push_back(d_term_util);
-
-  d_term_db = new quantifiers::TermDb( c, u, this );
-  d_util.push_back( d_term_db );
-  
   if (options::ceGuidedInst()) {
-    d_sygus_tdb = new quantifiers::TermDbSygus(c, this);
-  }else{
-    d_sygus_tdb = NULL;
+    d_sygus_tdb = std::unique_ptr<quantifiers::TermDbSygus>(
```
Comment (4tXJ7f) [inline src/theory/quantifiers_engine.cpp]: Nit: can be written as `d_sygus_tdb.reset(new quantifiers::TermDbSygus(c, this));`, which is more concise (this also applies to other places).

File: src/theory/quantifiers_engine.h [hunk 2]
```diff
@@ -103,28 +95,33 @@ class QuantifiersEngine {
   std::vector< QuantifiersUtil* > d_util;
   /** vector of modules for quantifiers */
   std::vector< QuantifiersModule* > d_modules;
+  //------------- quantifiers utilities
```
Comment (4tXJ7f) [inline src/theory/quantifiers_engine.h]: Maybe this is a good opportunity to move these private fields after the public ones?

File: src/theory/quantifiers_engine.h [hunk 3]
```diff
@@ -439,6 +313,117 @@ class QuantifiersEngine {
...
+  QuantifiersModule::QEffort d_curr_effort_level;
+  /** are we in conflict */
+  bool d_conflict;
+  context::CDO<bool> d_conflict_c;
+  /** has added lemma this round */
+  bool d_hasAddedLemma;
+  /** whether to use model equality engine */
+  bool d_useModelEe;
+  //------------- end temporary information during check
+ private:
```
[the remark this answers is not in the document]
Comment (ajreynol) [reply src/theory/quantifiers_engine.h]: At the moment, this is my marker for "below here needs reorganization".

[the remark this answers is not in the document]
Comment (4tXJ7f) [reply src/theory/quantifiers_engine.h]: Sounds good.

**Pull request merged**